### PR TITLE
Implement soft delete for Listings, Jobs

### PR DESCRIPTION
**Pull request by Kalsyc**
To do:

- Documentation
- Organisation endpoints

### CODE REVIEW

File: controllers/jobs.js [hunk 1]
```diff
@@ -163,6 +203,41 @@ exports.deleteJob = asyncHandler(async (req, res, next) => {
...
+  // check if job exists
+  const job = await db.oneOrNone(
+    'SELECT * FROM jobs WHERE job_id = $1',
+    req.params.id
+  );
+
+  // return bad request response if invalid job
+  if (!job) {
+    return next(new ErrorResponse(`Job does not exist`, 400));
+  }
```
Comment (don-tay) [inline controllers/jobs.js:232]: can be shortened to
```suggestion
  // check if job exists
  const job = await db.one(
    'SELECT * FROM jobs WHERE job_id = $1',
    req.params.id
  );
```

Convention is for `404` response if resource does not exist, which this method will do, so no need for `400` later on

File: controllers/jobs.js [hunk 2]
```diff
@@ -163,6 +203,41 @@ exports.deleteJob = asyncHandler(async (req, res, next) => {
     }
   }
 
+  const rows = await db.one(
+    'UPDATE jobs SET deleted_on=to_timestamp($2 / 1000.0) WHERE job_id = $1 RETURNING *',
+    [req.params.id, Date.now()]
+  );
```
Comment (don-tay) [inline controllers/jobs.js:209]: Lets standardise to using the `new Date().toLocaleString()` method from javascript to insert directly into PostgreSQL's timestamp fields

```suggestion
  const rows = await db.one(
    'UPDATE jobs SET deleted_on=$2 WHERE job_id = $1 RETURNING *',
    [req.params.id, new Date().toLocaleString()]
  );
```

File: controllers/jobs.js [hunk 3]
```diff
@@ -38,10 +67,21 @@ exports.getJobs = asyncHandler(async (req, res) => {
...
-  );
+
+  let rows;
+
+  if (req.query.all === 'true') {
+    rows = await db.one(
+      'SELECT * FROM jobs WHERE job_id = $1',
+      req.params.id
+    );
+  } else {
```
Comment (don-tay) [inline controllers/jobs.js:78]: Get single jobs no need handle `all`

File: controllers/jobs.js [hunk 4]
```diff
@@ -137,11 +177,11 @@ exports.updateJob = asyncHandler(async (req, res, next) => {
...
- * @access  Owner/Admin
+ * @desc    Deactivate (Soft delete) single job
+ * @route   PUT /api/jobs/:id/deactivate
+ * @access  Admin/Owner
  */
-exports.deleteJob = asyncHandler(async (req, res, next) => {
+exports.deactivateJob = asyncHandler(async (req, res, next) => {
   // check if job exists
   const job = await db.oneOrNone(
     'SELECT * FROM jobs WHERE job_id = $1',
```
Comment (don-tay) [inline controllers/jobs.js:187]:
```suggestion
  const job = await db.one(
    'SELECT * FROM jobs WHERE job_id = $1',
```
and remove the `400` response below it

File: controllers/jobs.js (hunk 1 above)
Comment (don-tay) [inline controllers/jobs.js:232]:
```suggestion
  // check if job exists
  const job = await db.one(
    'SELECT * FROM jobs WHERE job_id = $1',
    req.params.id
  );
    }
```

File: controllers/listings.js [hunk 5]
```diff
@@ -13,40 +13,92 @@ exports.getListings = asyncHandler(async (req, res) => {
...
+    if ((!req.user || req.user.role !== 'admin') && rows.deleted_on !== 'null') {
+      return next(new ErrorResponse(`Listing does not exist`, 400));
+
+    } else {
+      res.status(200).json({
+        success: true,
+        data: rows,
+      })
+    }
+
```
Comment (don-tay) [inline controllers/listings.js:51]: For `Get Single...`, no need to handle `all` query params. Frontend will decide on that.

File: controllers/listings.js [hunk 6]
```diff
@@ -13,40 +13,92 @@ exports.getListings = asyncHandler(async (req, res) => {
...
+  } else {
+    const rows = await db.one(
+      'SELECT * FROM listingsview WHERE listing_id = $1',
```
Comment (don-tay) [inline controllers/listings.js:54]: Following above comment:
```suggestion
      'SELECT * FROM listings WHERE listing_id = $1',
```

File: controllers/listings.js [hunk 7]
```diff
@@ -13,40 +13,92 @@ exports.getListings = asyncHandler(async (req, res) => {
...
-    data: rows,
-  });
+  if (req.query.all === 'true') {
+    const rows = await db.one(
+      'SELECT * FROM listings WHERE listing_id = $1',
+      decodedId
+    );
+    if ((!req.user || req.user.role !== 'admin') && rows.deleted_on !== 'null') {
+      return next(new ErrorResponse(`Listing does not exist`, 400));
+    } else {
```
Comment (don-tay) [inline controllers/listings.js:86]: `Get single...`  no need handle `all` query params

File: controllers/listings.js [hunk 8]
```diff
@@ -225,10 +277,41 @@ exports.verifyListing = asyncHandler(async (req, res, next) => {
...
+  // Unauthorised if neither admin nor listing owner
+  if (!(req.user.role === 'admin' || req.user.user_id === listing.created_by)) {
+    return next(
+      new ErrorResponse(`User not authorised to delete this listing`, 403)
+    );
+  }
+
+  const rows = await db.one(
+    'UPDATE listings SET deleted_on=to_timestamp($2 / 1000.0) WHERE listing_id = $1 RETURNING *',
+    [req.params.id, Date.now()]
```
Comment (don-tay) [inline controllers/listings.js:302]:
```suggestion
    'UPDATE listings SET deleted_on=$2 WHERE listing_id = $1 RETURNING *',
    [req.params.id, new Date().toLocaleString()]
```

File: middleware/advancedResults.js [hunk 9]
```diff
@@ -71,10 +71,12 @@ const advancedResults = (model, join, on) =>
     if (Object.keys(reqQuery).length !== 0) {
       filterQuery += 'WHERE ';
       for (let [key, value] of Object.entries(reqQuery)) {
+        
         const mappedFilter = {
           key,
           value: value.split(',').map(str => str.split(/'/).join('')), // remove all ' for pgp formatter to parse into sql
         };
+        console.log(mappedFilter)
```
Comment (don-tay) [inline middleware/advancedResults.js:79]: Remove `console.log` statement

File: controllers/listings.js [hunk 10]
```diff
@@ -13,40 +13,92 @@ exports.getListings = asyncHandler(async (req, res) => {
...
+exports.getListing = asyncHandler(async (req, res, next) => {
+
+  if (req.query.all === 'true') {
+    const rows = await db.one(
+      'SELECT * FROM listings WHERE listing_id = $1',
+      req.params.id
+    );
+
+    //if not admin and the listing has been deactivated, return a bad request
+    if ((!req.user || req.user.role !== 'admin') && rows.deleted_on !== 'null') {
```
Comment (don-tay) [inline controllers/listings.js:42]: use LoDash library

```
const  { get } = require('lodash')

const role = get(req, 'user.role', null);
```

File: middleware/auth.js [hunk 11]
```diff
@@ -3,6 +3,36 @@ const asyncHandler = require('./async');
...
+      decoded.id
+    );
+    next();
+    //if there is an error, there is no token or user is not authorized properly
+  } catch (err) {
+    next();
+  }
+
+})
+
```
Comment (don-tay) [inline middleware/auth.js:35]: No need, since `get single` endpoints not using this anymore

File: routes/job-routes.js [hunk 12]
```diff
@@ -12,14 +12,17 @@ const {
...
-  deleteJob
+  deleteJob,
+  deactivateJob,
 } = require('../controllers/jobs');
 
-router.route('/').get(advancedResults('jobs'), getJobs);
-router.route('/:id').get(getJob);
+router.route('/').get(advancedResults('jobsview'), getJobs);
+router.route('/all').get(protect, authorise('admin'), advancedResults('jobs'), getJobsAll);
+router.route('/:id').get(includeRoles, getJob);
```
Comment (don-tay) [inline routes/job-routes.js:25]: No need `includeRoles` middleware

File: routes/listing-routes.js [hunk 13]
```diff
@@ -75,8 +79,9 @@ router
     createListing
   );
 
-router.route('/:id/raw').get(getListing);
-router.route('/:hashId').get(getListingByHashId);
+router.route('/all').get(protect, authorise('admin'), advancedResults('listings'), getListingsAll);
+router.route('/:id/raw').get(includeRoles, getListing);
+router.route('/:hashId').get(includeRoles, getListingByHashId);
```
Comment (don-tay) [inline routes/listing-routes.js:84]: no need `includeRoles` middleware anymore

File: controllers/jobs.js [hunk 14]
```diff
@@ -163,6 +203,41 @@ exports.deleteJob = asyncHandler(async (req, res, next) => {
...
+  if (!job) {
+    return next(new ErrorResponse(`Job does not exist`, 400));
+  }
+
+  // if non-admin, check if owner of listing
+  if (req.user.role !== 'admin') {
+      return next(
+        new ErrorResponse(`Not authorised to delete jobs in this listing`, 403)
+        );
+    }
```
[the remark this answers is not in the document]
Comment (Kalsyc) [reply controllers/jobs.js:239]: Phasing out `authorise` middleware soon. Keeping the check for user roles in controllers for now.

File: controllers/jobs.js (hunk 1 above)
Comment (Kalsyc) [reply controllers/jobs.js:232]: Resolved.

File: controllers/jobs.js (hunk 2 above)
Comment (Kalsyc) [reply controllers/jobs.js:209]: Resolved.

File: controllers/jobs.js (hunk 3 above)
Comment (Kalsyc) [reply controllers/jobs.js:78]: Resolved.

File: controllers/listings.js [hunk 15]
```diff
@@ -298,8 +245,8 @@ exports.verifyListing = asyncHandler(async (req, res, next) => {
   }
 
   const rows = await db.one(
-    'UPDATE listings SET deleted_on=to_timestamp($2 / 1000.0) WHERE listing_id = $1 RETURNING *',
-    [req.params.id, Date.now()]
+    'UPDATE listings SET deleted_on=$2 WHERE listing_id = $1 RETURNING *',
+    [req.params.id, new Date().toLocaleDateString()]
```
Comment (don-tay) [inline controllers/listings.js:249]:
```suggestion
    [req.params.id, new Date().toLocaleString()]
```

**Comment by Kalsyc** [comment]
Okay resolved. This closes #48

**Pull request merged**